File: android/src/main/java/com/tencent/trtcplugin/render/VideoFrameDispatcher.java
package com.tencent.trtcplugin.render;

import android.util.SparseArray;

import com.tencent.trtc.TRTCCloudDef;
import com.tencent.trtc.TRTCCloudListener;
import com.tencent.trtcplugin.utils.TRTCLogger;

public class VideoFrameDispatcher implements TRTCCloudListener.TRTCVideoRenderListener {

    private static final String TAG = "VideoFrameDispatcher";
    private static final int MAX_STREAM_TYPES = 3;

    private final String mUserId;
    private final SparseArray<TextureRender> mRenders = new SparseArray<>(MAX_STREAM_TYPES);

    public VideoFrameDispatcher(String userId) {
        mUserId = userId;
    }

    public String getUserId() {
        return mUserId;
    }

    public synchronized void setRender(int streamType, TextureRender render) {
        TextureRender old = mRenders.get(streamType);
        if (old != null && old != render) {
            old.stop();
        }
        mRenders.put(streamType, render);
        TRTCLogger.i(TAG + " setRender userId=" + mUserId + " streamType=" + streamType);
    }

    public synchronized void removeRender(int streamType) {
        TextureRender render = mRenders.get(streamType);
        if (render != null) {
            render.stop();
            mRenders.remove(streamType);
            TRTCLogger.i(TAG + " removeRender userId=" + mUserId + " streamType=" + streamType);
        }
    }

    public synchronized boolean isEmpty() {
        return mRenders.size() == 0;
    }

    public synchronized void disposeAll() {
        for (int i = 0; i < mRenders.size(); i++) {
            mRenders.valueAt(i).stop();
        }
        mRenders.clear();
    }

    @Override
    public synchronized void onRenderVideoFrame(String userId, int streamType, TRTCCloudDef.TRTCVideoFrame frame) {
        TextureRender render = mRenders.get(streamType);
        if (render != null) {
            render.onVideoFrame(frame);
        }
    }
}

File: windows/view/video_frame_dispatcher.cpp
// Copyright (c) Tencent. All rights reserved.
//
// video_frame_dispatcher.cpp
//

#include "view/video_frame_dispatcher.h"

#include <string>

#include "view/texture_view_factory.h"

namespace trtc_sdk_flutter {

VideoFrameDispatcher::VideoFrameDispatcher(const std::string& userId)
    : user_id_(userId) {
}

VideoFrameDispatcher::~VideoFrameDispatcher() {
    disposeAll();
}

void VideoFrameDispatcher::setRender(int streamType, TextureRenderer* render) {
    std::lock_guard<std::mutex> lock(mutex_);
    renders_[streamType] = render;
}

void VideoFrameDispatcher::removeRender(int streamType) {
    std::lock_guard<std::mutex> lock(mutex_);
    renders_.erase(streamType);
}

void VideoFrameDispatcher::onRenderWillDispose(TextureRenderer* render) {
    if (render == nullptr) return;
    std::lock_guard<std::mutex> lock(mutex_);
    for (auto it = renders_.begin(); it != renders_.end();) {
        if (it->second == render) {
            it = renders_.erase(it);
        } else {
            ++it;
        }
    }
}

bool VideoFrameDispatcher::isEmpty() const {
    std::lock_guard<std::mutex> lock(mutex_);
    return renders_.empty();
}

void VideoFrameDispatcher::disposeAll() {
    std::lock_guard<std::mutex> lock(mutex_);
    renders_.clear();
}

void VideoFrameDispatcher::onRenderVideoFrame(const char* user_id, TRTCVideoStreamType stream_type,
    TRTCVideoFrame* frame) {
    std::lock_guard<std::mutex> lock(mutex_);
    auto it = renders_.find(static_cast<int>(stream_type));
    if (it != renders_.end() && it->second != nullptr) {
        it->second->onVideoFrame(frame);
    }
}

}  // namespace trtc_sdk_flutter
